# 회귀분석 11장 과제

> "작성 완료"

- toc: true
- branch: master
- badges: true
- comments: true
- author: 한재수
- categories: [R, statistics]

## 11.1

### 1)

In [1]:
F = 1803.30/0.78
F

[1] 2311.923

`-` $F$ 검정통계량이 $F(3,21;0.05)$ 보다 크므로 가정된 모형은 타당하다고 할 수 있다

### 2)

`-` $R^2=\dfrac{5409.89}{5426.24}=0.996986863832046$ : 종속변수의 변동중에서 설명변수들도 설명가능한 변동의 비율을 나타낸다, 설명변수들로 종속변수의 변동중 99%를 설명할 수 있다

`-` $R^2_{a}= 1-\left(\dfrac{n-1}{n-p-1}\right)(1-R^2)=0.996556415808052$ : 회귀모형에 추가된 독립변수가 유의할 때만 값이 증가하도록 만든 수정결정계수이다

`-` $MSE=0.78$ : 회귀모형 오차항의 $\sigma^2$의 추정량을 의미한다

`-` $C_p=\dfrac{1}{\hat{\sigma}^2}SSE_p-n+2(p+1)=3.96153846153846$ : 편향이 적은쪽으로 독립변수를 선택했을 때 $C_p$값이 $p+1$과 비슷해진다, $C_p$와 $p+1$이 비슷하므로 변수선택이 적절하다고 할 수 있다

## 11.5

In [183]:
x1 = c(10, 8, 12, 7, 12, 8, 12, 5, 8, 5, 11, 12, 6, 10, 10)
x2 = c(8, 6, 9, 16, 17, 15, 8, 10, 4, 16, 7, 12, 6, 4, 4)
x3 = c(5.5, 2.5, 8.0, 3.0, 2.9, 3.0, 8.0, 9.0, 4.0, 6.5, 5.5, 5.0, 6.0, 5.0, 3.5)
y = c(79, 200, 163, 200, 178, 146, 31, 292, 160, 339, 160, 86, 237, 107, 155)

### 1)

In [94]:
m = lm(y~x1+x2+x3)
m0 = lm(y~1)
anova(m0, m)

Res.Df RSS      Df Sum of Sq F        Pr(>F)    
1 14     89455.73 NA      NA         NA         NA
2 11     27853.53  3 61602.2   8.109376 0.00395365

`-` F 검정통계량에 근거한 p-value가 0.004로 0.05보다 작으므로 회귀모형은 유의하다 

### 2)

In [95]:
cor(x1, y); cor(x2, y); cor(x3, y)

[1] -0.7976095

[1] 0.3050202

[1] 0.1091833

`-` x1과 y의 상관관계가 제일 높으므로 y와 독립변수 x1만을 사용하여 회귀모형을 만든다 

In [96]:
m1 = lm(y~x1)
summary(m1)


Call:
lm(formula = y ~ x1)

Residuals:
    Min      1Q  Median      3Q     Max 
-66.518 -37.034  -8.585  30.449  82.516 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  395.686     49.296   8.027 2.16e-06 ***
x1           -25.017      5.247  -4.768 0.000367 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 50.04 on 13 degrees of freedom
Multiple R-squared:  0.6362,	Adjusted R-squared:  0.6082 
F-statistic: 22.73 on 1 and 13 DF,  p-value: 0.0003675


`-` F 검정통계량에 근거한 p-value가 매우 작으므로 추정된 회귀모형은 유의하다

In [97]:
m2 = lm(y~x1+x2); summary(m2)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-68.626 -31.714   6.395  37.622  66.819 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  354.219     58.101   6.097 5.36e-05 ***
x1           -24.206      5.164  -4.687 0.000526 ***
x2             3.604      2.826   1.275 0.226350    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 48.87 on 12 degrees of freedom
Multiple R-squared:  0.6796,	Adjusted R-squared:  0.6262 
F-statistic: 12.73 on 2 and 12 DF,  p-value: 0.001082


`-` 결정계수는 0.6796이다

In [98]:
m3 = lm(y~x1+x3); summary(m3)


Call:
lm(formula = y ~ x1 + x3)

Residuals:
    Min      1Q  Median      3Q     Max 
-72.134 -35.326  -9.401  26.031  88.013 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  381.367     63.243   6.030 5.94e-05 ***
x1           -24.905      5.436  -4.581 0.000631 ***
x3             2.578      6.733   0.383 0.708529    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 51.76 on 12 degrees of freedom
Multiple R-squared:  0.6406,	Adjusted R-squared:  0.5807 
F-statistic: 10.69 on 2 and 12 DF,  p-value: 0.002156


`-` 결정계수는 0.6406이다

`-` x1과 x2를 사용한 회귀모형의 결정계수가 x1과 x3를 사용한 회귀모형의 결정계수보다 크므로 x2를 추가할 독립변수로 선택한다

In [99]:
summary(m2)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-68.626 -31.714   6.395  37.622  66.819 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  354.219     58.101   6.097 5.36e-05 ***
x1           -24.206      5.164  -4.687 0.000526 ***
x2             3.604      2.826   1.275 0.226350    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 48.87 on 12 degrees of freedom
Multiple R-squared:  0.6796,	Adjusted R-squared:  0.6262 
F-statistic: 12.73 on 2 and 12 DF,  p-value: 0.001082


`-` 독립변수가 x2가 유의한지 검정하는 t통계량이 1.275이고 이에 대한 p-value가 0.05보다 크다

`-` 따라서 변수선택 절차는 여기서 끝나며 선택된 변수는 x1이고 회귀모형은 다음과 같다

`-` $\hat{y}=395.686-25.017x_1$

### 3)

In [100]:
m = lm(y~x1+x2+x3)
summary(m)


Call:
lm(formula = y ~ x1 + x2 + x3)

Residuals:
    Min      1Q  Median      3Q     Max 
-72.470 -28.485   4.682  39.100  59.125 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  330.940     72.646   4.555 0.000823 ***
x1           -23.993      5.331  -4.501 0.000900 ***
x2             3.826      2.936   1.303 0.219172    
x3             3.731      6.605   0.565 0.583528    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 50.32 on 11 degrees of freedom
Multiple R-squared:  0.6886,	Adjusted R-squared:  0.6037 
F-statistic: 8.109 on 3 and 11 DF,  p-value: 0.003954


`-` 위의 회귀모형에서 독립변수x3에 대한 t검정통계량에 근거한 p-value가 유의하지 않으면서 가장 크므로 x3를 회귀모형에서 제거한다

In [101]:
m = lm(y~x1+x2)
summary(m)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-68.626 -31.714   6.395  37.622  66.819 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  354.219     58.101   6.097 5.36e-05 ***
x1           -24.206      5.164  -4.687 0.000526 ***
x2             3.604      2.826   1.275 0.226350    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 48.87 on 12 degrees of freedom
Multiple R-squared:  0.6796,	Adjusted R-squared:  0.6262 
F-statistic: 12.73 on 2 and 12 DF,  p-value: 0.001082


`-` 위의 회귀모형에서 독립변수x2에 대한 t검정통계량에 근거한 p-value가 유의하지 않으면서 가장 크므로 x2를 회귀모형에서 제거한다

In [102]:
m = lm(y~x1)
summary(m)


Call:
lm(formula = y ~ x1)

Residuals:
    Min      1Q  Median      3Q     Max 
-66.518 -37.034  -8.585  30.449  82.516 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  395.686     49.296   8.027 2.16e-06 ***
x1           -25.017      5.247  -4.768 0.000367 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 50.04 on 13 degrees of freedom
Multiple R-squared:  0.6362,	Adjusted R-squared:  0.6082 
F-statistic: 22.73 on 1 and 13 DF,  p-value: 0.0003675


`-` 위의 회귀모형에서 독립변수x1에 대한 t검정통계량에 근거한 p-value가 유의하다

`-` 따라서 선택된 모형은 $\hat{y}=395.686-25.017x_1$이다

### 4)

In [103]:
cor(x1, y); cor(x2, y); cor(x3, y)

[1] -0.7976095

[1] 0.3050202

[1] 0.1091833

`-` x1과 y의 상관관계가 제일 높으므로 y와 독립변수 x1만을 사용하여 회귀모형을 만든다 

In [104]:
m1 = lm(y~x1)
summary(m1)


Call:
lm(formula = y ~ x1)

Residuals:
    Min      1Q  Median      3Q     Max 
-66.518 -37.034  -8.585  30.449  82.516 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  395.686     49.296   8.027 2.16e-06 ***
x1           -25.017      5.247  -4.768 0.000367 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 50.04 on 13 degrees of freedom
Multiple R-squared:  0.6362,	Adjusted R-squared:  0.6082 
F-statistic: 22.73 on 1 and 13 DF,  p-value: 0.0003675


`-` F 검정통계량에 근거한 p-value가 매우 작으므로 추정된 회귀모형은 유의하다

In [105]:
m2 = lm(y~x1+x2); summary(m2)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-68.626 -31.714   6.395  37.622  66.819 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  354.219     58.101   6.097 5.36e-05 ***
x1           -24.206      5.164  -4.687 0.000526 ***
x2             3.604      2.826   1.275 0.226350    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 48.87 on 12 degrees of freedom
Multiple R-squared:  0.6796,	Adjusted R-squared:  0.6262 
F-statistic: 12.73 on 2 and 12 DF,  p-value: 0.001082


`-` 결정계수는 0.6785이다

In [106]:
m3 = lm(y~x1+x3); summary(m3)


Call:
lm(formula = y ~ x1 + x3)

Residuals:
    Min      1Q  Median      3Q     Max 
-72.134 -35.326  -9.401  26.031  88.013 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  381.367     63.243   6.030 5.94e-05 ***
x1           -24.905      5.436  -4.581 0.000631 ***
x3             2.578      6.733   0.383 0.708529    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 51.76 on 12 degrees of freedom
Multiple R-squared:  0.6406,	Adjusted R-squared:  0.5807 
F-statistic: 10.69 on 2 and 12 DF,  p-value: 0.002156


`-` 결정계수는 0.6382이다

`-` x1과 x2를 사용한 회귀모형의 결정계수가 x1과 x3를 사용한 회귀모형의 결정계수보다 크므로 x2를 추가할 독립변수로 선택한다

In [107]:
summary(m2)


Call:
lm(formula = y ~ x1 + x2)

Residuals:
    Min      1Q  Median      3Q     Max 
-68.626 -31.714   6.395  37.622  66.819 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  354.219     58.101   6.097 5.36e-05 ***
x1           -24.206      5.164  -4.687 0.000526 ***
x2             3.604      2.826   1.275 0.226350    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 48.87 on 12 degrees of freedom
Multiple R-squared:  0.6796,	Adjusted R-squared:  0.6262 
F-statistic: 12.73 on 2 and 12 DF,  p-value: 0.001082


`-` 독립변수가 x2가 유의한지 검정하는 t통계량이 1.291이고 이에 대한 p-value가 0.05보다 크다

`-` 따라서 변수선택 절차는 여기서 끝나며 선택된 변수는 x1이고 회귀모형은 다음과 같다

`-` $\hat{y}=394.569-24.916x_1$

### 5)

In [122]:
m = lm(y~x1+x2+x3)
mse = anova(m0, m)[2,2] / 11
mse

[1] 2532.139

`-` $P=1$ 일 때

In [136]:
p = 1

In [137]:
anova(m0, lm(y~x1))[2,2] / mse - 15 + 2*(p+1)

[1] 1.853049

In [138]:
anova(m0, lm(y~x2))[2,2] / mse - 15 + 2*(p+1)

[1] 21.04129

In [139]:
anova(m0, lm(y~x3))[2,2] / mse - 15 + 2*(p+1)

[1] 23.90698

`-` $C_p$값이 $P+1$에 제일 가까운 $x_1$을 선택한다

`-` $P=2$ 일 때

In [140]:
p = 2

In [141]:
anova(m0, lm(y~x1+x2))[2,2] / mse - 15 + 2*(p+1)

[1] 2.319027

In [142]:
anova(m0, lm(y~x1+x3))[2,2] / mse - 15 + 2*(p+1)

[1] 3.697952

In [143]:
anova(m0, lm(y~x2+x3))[2,2] / mse - 15 + 2*(p+1)

[1] 22.25872

`-` $C_p$값이 $P+1$에 제일 가까운 $x_1, x_2$를 선택한다

`-` $P=3$ 일 때

In [144]:
p = 3

In [145]:
anova(m0, lm(y~x1+x2+x3))[2,2] / mse - 15 + 2*(p+1)

[1] 4

`-` $C_p$값이 $P+1$에 제일 가까운 $x_1, x_2, x_3$를 선택한다

### 6)

In [173]:
n = 15

`-` $P=1$일 때

In [168]:
p = 1

In [169]:
r = summary(lm(y~x1))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] 0.6081948

In [170]:
r = summary(lm(y~x2))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] 0.02327096

In [171]:
r = summary(lm(y~x3))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] -0.06408508

`-` $R_{ap}$가 제일큰 $x_1$을 선택한다

`-` $P=2$일 때

In [174]:
p = 2

In [175]:
r = summary(lm(y~x1+x2))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] 0.6262035

In [176]:
r = summary(lm(y~x1+x3))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] 0.5806662

In [177]:
r = summary(lm(y~x2+x3))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] -0.0322795

`-` $R_{ap}$가 제일큰 $x_1,x_2$를 선택한다

`-` $P=3$일 때

In [178]:
p = 3

In [182]:
r = summary(lm(y~x1+x2+x3))$r.squared
1 - (n-1)/(n-p-1)*(1-r)

[1] 0.6037152

`-` $R_{ap}$가 제일큰 $x_1,x_2,x_3$를 선택한다